nipype/pipeline/engine/workflows.py

```diff
@@ -903,21 +913,32 @@ def _set_node_input(self, node, param, source, sourceinfo):
         node.set_input(param, deepcopy(newval))
 
     def _get_all_nodes(self):
-        allnodes = []
-        for node in self._graph.nodes():
-            if isinstance(node, Workflow):
-                allnodes.extend(node._get_all_nodes())
-            else:
-                allnodes.append(node)
+        allnodes = [
+            *self._nodes_cache.difference(self._nested_workflows_cache)
+        ]  # all nodes that are not workflows
+        for node in self._nested_workflows_cache:
+            allnodes.extend(node._get_all_nodes())
         return allnodes
 
+    def _update_node_cache(self):
+        nodes = set(self._graph)
+
+        added_nodes = nodes.difference(self._nodes_cache)
+        removed_nodes = self._nodes_cache.difference(nodes)
+
+        self._nodes_cache = nodes
+        self._nested_workflows_cache.difference_update(removed_nodes)
+
+        for node in added_nodes:
+            if isinstance(node, Workflow):
+                self._nested_workflows_cache.add(node)
+
     def _has_node(self, wanted_node):
```

Comment:
This method is a depth-first search, and it seems that there should be ways to cut branches early and save effort. Does `wanted_node` have information we can use, such as `wanted_node.name.startswith(subworkflow.name)`? I can't really remember when and where the hierarchical names get attached.

Reply:
Okay, so we're literally just looking for "is this node object *somewhere* in this potentially hierarchical workflow"? So it's just following a tree of disjoint sets, which isn't the worst. Are we still spending a significant amount of time in `_has_node()` or are the cache updates a more worthwhile target?

The reason I ask is because it's not obvious we really need to recurse. Looking at each call to `_has_node()`, most make sense for the local workflow, and the recursive case seems to be covered by using `_get_all_nodes()`:

https://github.com/nipy/nipype/blob/0b49a70a007d9de7387b63db20dcea43fb263d49/nipype/pipeline/engine/workflows.py#L334-L344

Maybe there's a pathological construction strategy that we're heading off by recursing, so I wouldn't drop it without making sure we have tests for those strategies.

Reply:
> I'm not sure I follow what you mean by needing recursion

I think in >95% of use cases, `_has_node(node)` could be replaced with `node in self._node_cache` and not recursing down into sub-workflows. Something like fmriprep's style of doing:

```Python
def init_x_wf():
    wf = Workflow()
    wf.add_nodes([locally, defined, nodes])
    return wf
```

Ensures that nodes that are added to sub- or super-workflows aren't in-scope to be added. If we're doing an object-identity check, then there's no point in me checking sub-workflows for a node that the subworkflow can't know about.

> or pathological construction.

Here I mean a situation where you're making multiple workflows in the same scope and you actually could do:

```Python
wf1 = Workflow()
wf2 = Workflow()
bad_node = Node()
wf2.add_nodes([bad_node])
wf1.add_nodes([wf2])
wf1.add_nodes([bad_node])
```

I'm pretty sure this should fail now, and if we were to stop recursing in `_has_node()`, we'd need to make sure that this case is still causing an error by some other check.

Reply:
Also, my thought with the `_get_all_nodes()` check was:

```diff
  all_nodes = self._get_all_nodes()
  for node in nodes:
-     if self._has_node(node):
+     if node in all_nodes:
          raise IOError("Node %s already exists in the workflow" % node)
      if isinstance(node, Workflow):
          for subnode in node._get_all_nodes():
              if subnode in all_nodes:
                  raise IOError(
                      ("Subnode %s of node %s already exists " "in the workflow")
                      % (subnode, node)
                  )
```

This is a case where the `_has_node()` recursion is necessary, and we are already `_get_all_nodes()`ing, so that's why I thought we might be able to drop it.

Reply:
Thank you for the explanation! That makes sense. Personally, I would not worry too much about the recursion, the code is still quite efficient.

That's a good point, I can add that.